mapper/orm/src/main/java/org/hibernate/search/mapper/orm/tenancy/spi/TenancyConfiguration.java

```diff
@@ -34,41 +41,122 @@ public class TenancyConfiguration {
 					.validate( value -> Contracts.assertNotNullNorEmpty( value, "value" ) )
 					.build();
 
-	public static TenancyConfiguration create(TenancyMode tenancyMode,
-			ConfigurationPropertySource configurationPropertySource) {
+	public static TenancyConfiguration create(TenancyMode tenancyMode, ConfigurationPropertySource configurationPropertySource,
+			BeanResolver beanResolver) {
 		String tenantIdsConfigurationPropertyKey = MULTI_TENANCY_TENANT_IDS.resolveOrRaw( configurationPropertySource );
 		switch ( tenancyMode ) {
 			case SINGLE_TENANCY:
-				return new TenancyConfiguration( Optional.of( Collections.emptySet() ),
-						tenantIdsConfigurationPropertyKey );
+				return new Single();
 			case MULTI_TENANCY:
-				return new TenancyConfiguration(
-						MULTI_TENANCY_TENANT_IDS.getAndMap( configurationPropertySource, LinkedHashSet::new ),
-						tenantIdsConfigurationPropertyKey );
+				// get from property:
+				BeanReference<SimpleToStringTenantIdentifierConverter> converterReference =
+						BeanReference.ofInstance( new SimpleToStringTenantIdentifierConverter() );
+				BeanHolder<? extends TenantIdentifierConverter> converterBeanHolder =
+						beanResolver.resolve( converterReference );
+				return MULTI_TENANCY_TENANT_IDS.get( configurationPropertySource )
+						.map( tenants -> (TenancyConfiguration) new MultiConfigured(
+								converterBeanHolder,
+								tenants,
+								tenantIdsConfigurationPropertyKey
+						) ).orElseGet( () -> new MultiFailing( converterBeanHolder, tenantIdsConfigurationPropertyKey ) );
 		}
 		throw new AssertionFailure( "Unknown tenancy mode: " + tenancyMode );
 	}
 
-	private final Optional<Set<String>> tenantIds;
-	private final String tenantIdsConfigurationPropertyKey;
+	private static final class Single extends TenancyConfiguration {
 
-	private TenancyConfiguration(Optional<Set<String>> tenantIds, String tenantIdsConfigurationPropertyKey) {
-		this.tenantIds = tenantIds;
-		this.tenantIdsConfigurationPropertyKey = tenantIdsConfigurationPropertyKey;
+		@Override
+		public Set<TenantIdentifier<?>> tenantIdsOrFail() {
+			return Collections.emptySet();
+		}
+
+		@Override
+		public SearchException invalidTenantId(TenantIdentifier<?> tenantId) {
+			throw new AssertionFailure( "Single tenancy is used." );
+		}
+
+		@Override
+		public TenantIdentifier<?> convert(Object tenantIds) {
+			return TenantIdentifier.NO_TENANT;
```

Comment:
This is a good idea, but I wonder how many `tenantIdentifier() != null` we have out there... we'll need to catch them all and replace them with `tenantIdentifier().isEmpty()` or something like that, otherwise we'll introduce many bugs for people using single-tenancy, which is most people.